src/Common/ThreadPool.cpp

```diff
@@ -359,18 +350,39 @@ void ThreadPoolImpl<Thread>::worker(typename std::list<Thread>::iterator thread_
 
         /// Get a job from the queue.
         Job job;
-        std::exception_ptr exception_from_job;
-        bool need_shutdown = false;
 
         {
             std::unique_lock lock(mutex);
-            new_job_or_shutdown.wait(lock, [&] { return !jobs.empty() || shutdown || (threads.size() > std::min(max_threads, scheduled_jobs + max_free_threads)); });
-            need_shutdown = shutdown;
 
-            if (jobs.empty())
+            // Finish with previous job if any
+            if (job_is_done)
+            {
+                job_is_done = false;
+                if (exception_from_job)
+                {
+                    if (!first_exception)
+                        first_exception = exception_from_job;
+                    if (shutdown_on_exception)
+                        shutdown = true;
+                    exception_from_job = {};
+                }
+
+                --scheduled_jobs;
+
+                job_finished.notify_all();
+                if (shutdown)
+                    new_job_or_shutdown.notify_all(); /// `shutdown` was set, wake up other threads so they can finish themselves.
+            }
+
+            new_job_or_shutdown.wait(lock, [&] { return !jobs.empty() || shutdown || threads.size() > std::min(max_threads, scheduled_jobs + max_free_threads); });
+
+            if (jobs.empty() || threads.size() > std::min(max_threads, scheduled_jobs + max_free_threads))
             {
-                /// No jobs and either `shutdown` is set or this thread is excessive. The worker will stop.
```

Comment:
This comment can be useful

Reply:
Actually, it is wrong and I spent some time following it blindly. Right two conditions are:
1) worker is not needed anymore (exceed max free worker count).
2) shutdown + no more jobs. Because we want to handle every job in the queue on shutdown, so shut down is handled below separately.
Anyways, a comment is required. I'll add it.